### ISSUE TITLE
S1226: rule doesn't consider null checking as a read

### ISSUE THREAD

**Issue by Elanis** [SonarSource/sonar-dotnet#2603]
Hello,

This issue is about rule **S1226**: *Method parameters, caught exceptions and foreach variables' initial values should not be ignored*.

When we run analysis on our projects with this kind of code:
```cs
public ReturnObject Foo(MyParamObject bar) {
    if (bar == null)
         bar = new MyParamObject(); // Sonar raise a S1226 bug here
    }

    // ...
}
```

Sonar detect it as a bug, but the variable is read (null checking). Is it a false positive ?

**Sonarqube version:** 7.9.1.27448
**Sonar C# version:** 7.16 (build 8981)
**SonarScanner for Team Foundation Server version:** 4.7.2
**SonarScanner-Msbuild version:** 4.6.2

Thanks :)

**Comment by Corniel** [comment]
I think Sonar is ride here. You should do something like:

``` C#
public ReturnObject Foo(MyParamObject bar) {
    var b = bar ?? new MyParamObject();
    // ...
}
```

Just as suggested, you should not reuse the parameter, but instead introduce a new variable.

**Comment by Elanis** [comment]
The rule description says

> While it is technically correct to assign to parameters from within method bodies, doing so before the parameter value is read is likely a bug. Instead, initial values of parameters, caught exceptions, and foreach parameters should be, if not treated as final, then at least read before reassignment.

I read this as "Read the parameter at least one time, then you can reuse/overwrite it". Am I wrong ?

**Comment by neolution-ch** [comment]
#2555 is related to this

**Comment by Corniel** [comment]
@Elanis : I assume (always tricky) that the implementation triggers the rule when a reassign is done, no matter if is validated. The description is therefore confusing/not matching the behaviour. I would argue though, that you should never reassign a parameter, but introduce a new one, if you want to change the value before proceeding, just as I suggested.

But let us wait what the Sonar guys/girls think about this.

**Comment by costin-zaharia-sonarsource** [comment]
It is indeed a false positive. I'm closing this issue as a duplicate of #2555.

**Comment by Elanis** [comment]
Thanks for your answer :)
